File: gps/utilities/defines_parser/src/parser.cpp
#include "parser.h"

using namespace std;

TreeNode* Parser::Parse() throw(ParserError)
{
    TreeNode *expression=ParseExpression();

    if(tokenizer.HasNext())
    {
        delete expression;
        throw ParserError("characters remaining in expression");
    }
    
    return expression;
}

/**
* Parse next expression.
* Expressions are defined as: Sum [ : Sum | '[hdb] Expression ]
*/
TreeNode* Parser::ParseExpression() throw(SyntaxError)
{
    //Read range.
    TreeNode *expression=ParseSum();

    try
    {
        if(tokenizer.HasNext())
        {
            if(tokenizer.NextType()==TokenType::COLON)
            {
                tokenizer.ReadNext();
                expression=new TreeNode(TokenType::COLON,expression,NULL);

                TreeNode *right;
                try{ right=ParseSum(); }//Get second term
                catch(...){ delete expression; throw; }
                expression->SetRight(right);
            }
            else if(tokenizer.NextType()==TokenType::CONST)
            {
                expression=new TreeNode(TokenType::CONST,expression,NULL);
                expression->SetValue(tokenizer.ReadNext());
        
                TreeNode *right;
                try{ right=ParseSum(); }//Get second term
                catch(...){ delete expression; throw; }
                expression->SetRight(right);
            }
        }
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
    
    return expression;
}

/**
* Parse next sum.
* Sums are defined as: [-] Term { (+ | -) Term }
*/
TreeNode* Parser::ParseSum() throw(SyntaxError)
{
    try
    {
        //Read initial minus sign.
        bool minus=(tokenizer.NextType()==TokenType::MINUS);
        if(minus)tokenizer.ReadNext();

        TreeNode *expression=ParseTerm();//Get term
    
        if(minus)expression=new TreeNode(TokenType::MINUS,new TreeNode(TokenType::NUMBER,"0"),expression);//Make negative
    
        while(tokenizer.HasNext() &&
              (tokenizer.NextType()==TokenType::PLUS ||
               tokenizer.NextType()==TokenType::MINUS))
        {
            TreeNode *t=new TreeNode(tokenizer.NextType(),NULL,NULL);
            t->SetLeft(expression);
            expression=t;
            tokenizer.ReadNext();//Eat sign
            TreeNode *right;
            try{ right=ParseTerm(); }//Get second term
            catch(...){ delete expression; throw; }
            expression->SetRight(right);//Set right
        }
        
        return expression;
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

/**
* Parse next term.
* Terms are defined as: Factor { ( * | / ) Factor }
*/
TreeNode* Parser::ParseTerm() throw(SyntaxError)
{
    try
    {
        TreeNode *term=ParseFactor();//Get factor
    
        while(tokenizer.HasNext() &&
              (tokenizer.NextType()==TokenType::TIMES ||
               tokenizer.NextType()==TokenType::DIVIDE))
        {
            term=new TreeNode(tokenizer.NextType(),term,NULL);//Create new node and move old term left
            tokenizer.ReadNext();//Eat sign
            TreeNode *right;
            try{ right=ParseFactor(); }//Get second operand
            catch(...){ delete term; throw; }
            term->SetRight(right);//Set right
        }
    
        return term;//Return term
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

/**
* Parse next factor.
* Factors are defined as: Base [ ^ Sum ]
*/
TreeNode* Parser::ParseFactor() throw(SyntaxError)
{
    try
    {
        TreeNode *base=ParseBase();//Get base
        if(tokenizer.HasNext() &&
           tokenizer.NextType()==TokenType::CARET)
        {
            tokenizer.ReadNext();//Eat carrot
            bool minus=(tokenizer.NextType()==TokenType::MINUS);//Minus sign?
            if(minus)tokenizer.ReadNext();//Eat minus
            TreeNode *exponent;
            try{ exponent=ParseSum(); }//Get number
            catch(...){ delete base; throw; }
            
            if(minus)
            {
                exponent=new TreeNode(TokenType::MINUS,
                                      new TreeNode(TokenType::NUMBER,"0"),
                                      exponent);//Make negative
            }
            return new TreeNode(TokenType::CARET,base,exponent);//Return factor
        }
        else return base;//Return base alone
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

/**
* Parse next base.
* Bases are defined as:
* Number | Variable | HEX | Function ( Sum )
*/
TreeNode* Parser::ParseBase() throw(SyntaxError)
{
    try
    {
        switch(tokenizer.NextType()){
        case TokenType::NUMBER: return ParseValue();
        case TokenType::VARIABLE: return ParseVariable();
        case TokenType::HEX: return ParseHex();
        case TokenType::FUNCTION: return ParseFunction();
        default:
            if(tokenizer.NextType()!=TokenType::LPAREN)
            {
                throw SyntaxError("expected left paren '('");
            }
            tokenizer.ReadNext();//Eat paren
            TreeNode *expression=ParseSum();
            if(tokenizer.NextType()!=TokenType::RPAREN)
            {
                delete expression;
                throw SyntaxError("expected right paren ')'");
            }
            tokenizer.ReadNext();//Eat paren
            return expression;
        }
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

/**
* Parse next function.
* Functions are defined as: Function ( Sum )
*/
TreeNode* Parser::ParseFunction() throw(SyntaxError)
{
    try
    {
        if(tokenizer.NextType()!=TokenType::FUNCTION)throw SyntaxError("expected function");

        string functionName=tokenizer.ReadNext();
        TreeNode *function=new TreeNode(TokenType::FUNCTION,functionName);//Parse and return leaf (value 0 for non-leaf)

        //Invalid syntax - missing paren
        if(!tokenizer.HasNext() ||
           tokenizer.NextType()!=TokenType::LPAREN)
        {
            delete function;
            throw SyntaxError("expected left paren '('");
        }
        tokenizer.ReadNext();//Eat paren
    
        TreeNode *expression;
        try{ expression=ParseSum(); }
        catch(...){ delete function; throw; }
        function->SetRight(expression);

        //Invalid syntax - missing paren
        if(!tokenizer.HasNext() ||
           tokenizer.NextType()!=TokenType::RPAREN)
        {
            delete function;
            delete expression;
            throw SyntaxError("expected right paren ')'");
        }
        tokenizer.ReadNext();//Eat paren
    
        return function;//Return function
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

TreeNode* Parser::ParseValue() throw(SyntaxError)
{
    try
    {
        if(tokenizer.NextType()!=TokenType::VALUE)throw SyntaxError("expected value");
        return new TreeNode(TokenType::VALUE,tokenizer.ReadNext());
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

TreeNode* Parser::ParseVariable() throw(SyntaxError)
{
    try
    {
        if(tokenizer.NextType()!=TokenType::VARIABLE)throw SyntaxError("expected variable");
        return new TreeNode(TokenType::VARIABLE,tokenizer.ReadNext());
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

TreeNode* Parser::ParseHex() throw(SyntaxError)
{
    try
    {
        if(tokenizer.NextType()!=TokenType::HEX)throw SyntaxError("expected value");
        return new TreeNode(TokenType::HEX,tokenizer.ReadNext());
    }
    catch(Tokenizer::UnknownTokenException &e){ throw SyntaxError("unknown token '"+e.GetToken()+"'"); }
    catch(Tokenizer::OutOfBoundsException){ throw(SyntaxError("unexpected end of expression")); }
}

File: gps/utilities/defines_parser_java/Parser.java
import Exceptions.ParserError;
import Exceptions.SyntaxError;
import Exceptions.UnknownTokenException;

public class Parser
{
    Tokenizer tokenizer;
    
    public Parser(String expression)
    {
        tokenizer=new Tokenizer(expression);
    }
    
    public static TreeNode Parse(String expression) throws ParserError
    {
        return (new Parser(expression)).Parse();
    }

    public TreeNode Parse() throws ParserError
    {
        TreeNode expression=ParseExpression();
        if(tokenizer.HasNext())throw new ParserError("characters remaining in expression");
        return expression;
    }

    /**
    * Parse next expression.
    * Expressions are defined as: @ | Sum [ : Sum | '[hdb] Expression ]
    */
    public TreeNode ParseExpression() throws SyntaxError
    {
        try
        {
            if(tokenizer.HasNext() && tokenizer.NextType()==Tokenizer.TokenType.AT)
            {
                tokenizer.ReadNext();
                return new TreeNode(Tokenizer.TokenType.AT,null,null);
            }
        }
        catch(UnknownTokenException e)
        {
            throw new SyntaxError(e.getMessage());
        }
        
        //Read range.
        TreeNode expression=ParseSum();

        try
        {
            if(tokenizer.HasNext())
            {
                if(tokenizer.NextType()==Tokenizer.TokenType.COLON)
                {
                    tokenizer.ReadNext();
                    expression=new TreeNode(Tokenizer.TokenType.COLON,expression,null);

                    TreeNode right;
                    right=ParseSum();//Get second term
                    expression.SetRight(right);
                }
                else if(tokenizer.NextType()==Tokenizer.TokenType.CONST)
                {
                    expression=new TreeNode(Tokenizer.TokenType.CONST,expression,null);
                    expression.SetValue(tokenizer.ReadNext());
            
                    TreeNode right;
                    right=ParseSum();//Get second term
                    expression.SetRight(right);
                }
            }
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
        
        return expression;
    }

    /**
    * Parse next sum.
    * Sums are defined as: [-] Term { (+ | -) Term }
    */
    public TreeNode ParseSum() throws SyntaxError
    {
        try
        {
            //Read initial minus sign.
            boolean minus=(tokenizer.NextType()==Tokenizer.TokenType.MINUS);
            if(minus)tokenizer.ReadNext();

            TreeNode expression=ParseTerm();//Get term
        
            if(minus)expression=new TreeNode(Tokenizer.TokenType.MINUS,new TreeNode(Tokenizer.TokenType.NUMBER,"0"),expression);//Make negative
        
            while(tokenizer.HasNext() &&
                  (tokenizer.NextType()==Tokenizer.TokenType.PLUS ||
                   tokenizer.NextType()==Tokenizer.TokenType.MINUS))
            {
                TreeNode t=new TreeNode(tokenizer.NextType(),null,null);
                t.SetLeft(expression);
                expression=t;
                tokenizer.ReadNext();//Eat sign
                TreeNode right;
                right=ParseTerm();//Get second term
                expression.SetRight(right);
            }
            
            return expression;
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    /**
    * Parse next term.
    * Terms are defined as: Factor { ( * | / ) Factor }
    */
    public TreeNode ParseTerm() throws SyntaxError
    {
        try
        {
            TreeNode term=ParseFactor();//Get factor
        
            while(tokenizer.HasNext() &&
                  (tokenizer.NextType()==Tokenizer.TokenType.TIMES ||
                   tokenizer.NextType()==Tokenizer.TokenType.DIVIDE))
            {
                term=new TreeNode(tokenizer.NextType(),term,null);//Create new node and move old term left
                tokenizer.ReadNext();//Eat sign
                TreeNode right;
                right=ParseFactor();//Get second operand
                term.SetRight(right);//Set right
            }
        
            return term;//Return term
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    /**
    * Parse next factor.
    * Factors are defined as: Base [ ^ Sum ]
    */
    public TreeNode ParseFactor() throws SyntaxError
    {
        try
        {
            TreeNode base=ParseBase();//Get base
            if(tokenizer.HasNext() &&
               tokenizer.NextType()==Tokenizer.TokenType.CARET)
            {
                tokenizer.ReadNext();//Eat carrot
                boolean minus=(tokenizer.NextType()==Tokenizer.TokenType.MINUS);//Minus sign?
                if(minus)tokenizer.ReadNext();//Eat minus
                TreeNode exponent;
                exponent=ParseSum();//Get number
                
                if(minus)
                {
                    exponent=new TreeNode(Tokenizer.TokenType.MINUS,
                                          new TreeNode(Tokenizer.TokenType.NUMBER,"0"),
                                          exponent);//Make negative
                }
                return new TreeNode(Tokenizer.TokenType.CARET,base,exponent);//Return factor
            }
            else return base;//Return base alone
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    /**
    * Parse next base.
    * Bases are defined as:
    * Number | Variable | HEX | Function ( Sum [; Sum]... )
    */
    public TreeNode ParseBase() throws SyntaxError
    {
        try
        {
            switch(tokenizer.NextType()){
            case Tokenizer.TokenType.NUMBER: return ParseValue();
            case Tokenizer.TokenType.VARIABLE: return ParseVariable();
            case Tokenizer.TokenType.HEX: return ParseHex();
            case Tokenizer.TokenType.FUNCTION: return ParseFunction();
            default:
                if(tokenizer.NextType()!=Tokenizer.TokenType.LPAREN)
                {
                    throw new SyntaxError("expected left paren '('");
                }
                tokenizer.ReadNext();//Eat paren
                TreeNode expression=ParseSum();
                if(tokenizer.NextType()!=Tokenizer.TokenType.RPAREN)
                {
                    throw new SyntaxError("expected right paren ')'");
                }
                tokenizer.ReadNext();//Eat paren
                return expression;
            }
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    /**
    * Parse next function.
    * Functions are defined as: Function ( Sum [; Sum]... )
    */
    public TreeNode ParseFunction() throws SyntaxError
    {
        try
        {
            if(tokenizer.NextType()!=Tokenizer.TokenType.FUNCTION)throw new SyntaxError("expected function");

            String functionName=tokenizer.ReadNext();
            TreeNode function=new TreeNode(Tokenizer.TokenType.FUNCTION,functionName);//Parse and return leaf (value 0 for non-leaf)

            //Invalid syntax - missing paren
            if(!tokenizer.HasNext() ||
               tokenizer.NextType()!=Tokenizer.TokenType.LPAREN)
            {
                throw new SyntaxError("expected left paren '('");
            }
            tokenizer.ReadNext();//Eat paren
        
            TreeNode expression;
            expression=ParseSum();
            function.SetRight(expression);

            TreeNode parent=function;
            while(tokenizer.HasNext() &&
                  (tokenizer.NextType()==Tokenizer.TokenType.SEMICOLON))
            {
                tokenizer.ReadNext();
                parent.SetRight(new TreeNode(Tokenizer.TokenType.SEMICOLON,
                                             parent.GetRight(),
                                             null));
                parent=parent.GetRight();
                parent.SetRight(ParseSum());
            }

            //Invalid syntax - missing paren
            if(!tokenizer.HasNext() ||
               tokenizer.NextType()!=Tokenizer.TokenType.RPAREN)
            {
                throw new SyntaxError("expected right paren ')'");
            }
            tokenizer.ReadNext();//Eat paren
        
            return function;//Return function
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    public TreeNode ParseValue() throws SyntaxError
    {
        try
        {
            if(tokenizer.NextType()!=Tokenizer.TokenType.VALUE)throw new SyntaxError("expected value");
            return new TreeNode(Tokenizer.TokenType.VALUE,tokenizer.ReadNext());
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    public TreeNode ParseVariable() throws SyntaxError
    {
        try
        {
            if(tokenizer.NextType()!=Tokenizer.TokenType.VARIABLE)throw new SyntaxError("expected variable");
            return new TreeNode(Tokenizer.TokenType.VARIABLE,tokenizer.ReadNext());
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }

    public TreeNode ParseHex() throws SyntaxError
    {
        try
        {
            if(tokenizer.NextType()!=Tokenizer.TokenType.HEX)throw new SyntaxError("expected value");
            return new TreeNode(Tokenizer.TokenType.HEX,tokenizer.ReadNext());
        }
        catch(UnknownTokenException e){ throw new SyntaxError("unknown token '"+e.GetToken()+"'"); }
        catch(IndexOutOfBoundsException e){ throw new SyntaxError("unexpected end of expression"); }
    }
}